### PR TITLE
Create our own logic for chunking, content framing

### PR DESCRIPTION
**Pull request by sethmlarson**
- Changes `HTTPConnection.request_chunked()` into an alias for `HTTPConnection.request(..., chunked=True)`
- Adds a `chunked` parameter to `HTTPConnection.request()`
- Adds a new internal function `body_to_chunks()` which does discovery of Content-Length if applicable for the body/method and converts the body if any into an iterable of `socket.sendall()`-able chunks.
- Changes `HTTPConnection.request()` to use `.putrequest()`, `putheader()`, and `.endheaders()` and to send the body via `.send()` in chunks which was previously handled by `http.client.HTTPConnection`.

All of these changes are necessary for us to one day add another HTTP implementation as we're not relying on `http.client` logic for figuring out content framing, sending of data.

Closes https://github.com/urllib3/urllib3/issues/2515
Closes https://github.com/urllib3/urllib3/issues/2291

### CODE REVIEW

**Comment by pquentin** [comment]
This is exciting! Can you please look into the CI failures?

**Comment by sethmlarson** [comment]
We may need to work on #2242 before we can land this one now that we're sending headers without `.itermerged()`. Only failing test is:

```
test_headers_http_header_dict failed (1 runs remaining out of 2).
	<class 'AssertionError'>
	assert '1,2' == '1, 2'
  - 1, 2
  ?   -
  + 1,2
	[<TracebackEntry /Users/runner/work/urllib3/urllib3/test/with_dummyserver/test_poolmanager.py:357>]
test_headers_http_header_dict failed; it passed 0 out of the required 1 times.
	<class 'AssertionError'>
	assert '1,2' == '1, 2'
  - 1, 2
  ?   -
  + 1,2
	[<TracebackEntry /Users/runner/work/urllib3/urllib3/test/with_dummyserver/test_poolmanager.py:357>]
```

**Comment by pquentin** [comment]
I would be happy to just change the test to assert for `1,2` until we land #2242, but I am probably missing something. Also, do you understand the botocore failures?

**Comment by sethmlarson** [comment]
Discovered that the test was "failing" because we were relying on `tornado.httputil.HTTPHeaders` method of wrapping headers which uses `,` instead of `, `. I've added a new dummyserver handler which keeps the order that headers are received and doesn't fold header values so we can test our header behavior easier.

**Comment by sethmlarson** [comment]
Also I'm unsure about the botocore failures, going to ping our resident Amazonian @nateprewitt to see if he knows anything about these tests :)

**Comment by nateprewitt** [comment]
I didn't have a chance to deep dive what specifically is failing but it's definitely around our HTTP 100 functionality. We have some changes injected to handle responses for 100 Continue that are relying on functions in urllib3. It seems we're asserting some of them are called that aren't actually used anymore. I'll see if I can get some time next week to figure out if they're actually legitimate or poorly written tests.

**Comment by sethmlarson** [comment]
@nateprewitt I'm seeing https://github.com/boto/botocore/blob/8ea47e0ab9f505cc547ac19ce689fd20db3e0241/botocore/awsrequest.py#L109 which appears to add support for `Expect: 100-Continue`, what are your thoughts on upstreaming that logic to urllib3?

**Comment by nateprewitt** [comment]
We can look at that, but I believe the boto implementation is essentially a hack. It doesn't perform to spec, just enough to be interoperable with s3.

I can look at putting together a test suite from the rfc to validate behavior, but I recall Cory saying it wasn't possible with httplib.

**Comment by nateprewitt** [comment]
Alrighty, I think this is going to take a while for us to unravel in Botocore. I spoke with Seth this weekend at PyCon and I think the best path forward for now is to skip the failing test in Botocore and we’ll track the fix on our side.

That will get this merged so the prep for 2.0 can move forward and I’ll follow up when it’s safe to remove.

**Review by nateprewitt — APPROVED** [review]

**Review by pquentin — CHANGES_REQUESTED** [review]
Thanks! I think we're nearly there.

File: test/with_dummyserver/test_poolmanager.py [hunk 1]
```diff
@@ -351,26 +351,31 @@ def test_headers_http_header_dict(self) -> None:
...
 
         with PoolManager(headers=headers) as http:
-            r = http.request("GET", f"{self.base_url}/headers")
-            returned_headers = r.json()
-            assert returned_headers["Foo"] == "bar"
-            assert returned_headers["Multi"] == "1, 2"
-            assert returned_headers["Baz"] == "quux"
+            r = http.request("GET", f"{self.base_url}/multi_headers")
+            returned_headers = r.json()["headers"]
+            assert returned_headers[-4:] == [
```
[the remark this answers is not in the document]
Comment (sethmlarson) [reply test/with_dummyserver/test_poolmanager.py:356]: We're using a list here to maintain order, I can add a comment about this so we don't lose the coverage there.

File: changelog/2515.feature.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Added the ``chunked`` parameter to ``urllib3.connection.HTTPConnection.request()``.
+
+Changed ``urllib3.connection.HTTPConnection.request_chunked()`` to be an alias for `HTTPConnection.request(..., chunked=True)``.
+
+Changed ``urllib3.connection.HTTPConnection.request()`` to use ``http.client.HTTPConnection.putrequest()``, ``putheader()``, ``.endheaders()``, and ``.send()`` methods directly.
```
[the remark this answers is not in the document]
Comment (sethmlarson) [reply changelog/2515.feature.rst:5]: Sounds good to me! Thanks for calling out implementation details leaking into changelogs.

**Comment by pquentin** [comment]
This now looks good, but the Windows failures are very mysterious but seem related to the change.

**Comment by sethmlarson** [comment]
@pquentin I've removed the wrapping of `request_chunked` in our HTTPConnection marker class. Not sure if this fixed the Windows issue we were seeing but things are passing now? Either way we don't want to wrap `request_chunked` there to avoid double sending of the marker.

**Review by pquentin — APPROVED** [review]
LGTM! Thanks for iterating.

**Pull request merged**